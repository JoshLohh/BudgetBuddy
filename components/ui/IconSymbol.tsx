// Fallback for using MaterialIcons on Android and web.

import MaterialIcons from '@expo/vector-icons/MaterialIcons';
import { SymbolWeight, SymbolViewProps } from 'expo-symbols';
import { ComponentProps } from 'react';
import { OpaqueColorValue, type StyleProp, type TextStyle } from 'react-native';
import { View } from 'react-native';

type IconMapping = typeof MAPPING;
//type IconMapping = Record<SymbolViewProps['name'], ComponentProps<typeof MaterialIcons>['name']>;
//type IconSymbolName = keyof typeof MAPPING;
type IconSymbolName = keyof IconMapping;

/**
 * Add your SF Symbols to Material Icons mappings here.
 * - see Material Icons in the [Icons Directory](https://icons.expo.fyi).
 * - see SF Symbols in the [SF Symbols](https://developer.apple.com/sf-symbols/) app.
 */
const MAPPING = {
  'house.fill': 'home',
  'paperplane.fill': 'send',
  'chevron.left.forwardslash.chevron.right': 'code',
  'chevron.right': 'chevron-right',
  'plus.square.fill.on.square.fill': 'square',
  'person.3.fill': 'groups',
  'person.fill': 'add',

} as const;

/**
 * An icon component that uses native SF Symbols on iOS, and Material Icons on Android and web.
 * This ensures a consistent look across platforms, and optimal resource usage.
 * Icon `name`s are based on SF Symbols and require manual mapping to Material Icons.
 */
export function IconSymbol({
  name,
  size = 24,
  color,
  style,
}: {
  name: IconSymbolName;
  size?: number;
  color: string | OpaqueColorValue;
  style?: StyleProp<TextStyle>;
  weight?: SymbolWeight;
}) {
  //return <MaterialIcons testID="material-icon" color={color} size={size} name={MAPPING[name]} style={style} />;
  return (
    <MaterialIcons
      testID="material-icon"  // Move testID here
      color={color}
      size={size}
      name={MAPPING[name]}
      style={style}
    />
  );
}
