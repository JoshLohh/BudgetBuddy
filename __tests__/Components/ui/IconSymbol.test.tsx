import React from 'react';
import { render } from '@testing-library/react-native';
import { IconSymbol } from '../../../components/ui/IconSymbol.tsx';
import MaterialIcons from '@expo/vector-icons/MaterialIcons';

// Mock MaterialIcons to spy on its calls
jest.mock('@expo/vector-icons/MaterialIcons', () => {
  return jest.fn(() => null);
});

describe('IconSymbol tests using toHaveBeenCalledWith pattern', () => {
  beforeEach(() => {
    (MaterialIcons as unknown as jest.Mock).mockClear();
  });

  it('calls MaterialIcons with correct mapped names and props for house.fill', () => {
    render(<IconSymbol name="house.fill" color="#fff" size={24} />);
    expect(MaterialIcons).toHaveBeenCalledWith(
      expect.objectContaining({
        name: 'home',
        size: 24,
        style: undefined,
        color: '#fff'
      }),
      undefined
    );
  });

  it('calls MaterialIcons with correct mapped names and custom size', () => {
    render(<IconSymbol name="paperplane.fill" color="#000" size={30} />);
    expect(MaterialIcons).toHaveBeenCalledWith(
      expect.objectContaining({
        name: 'send',
        size: 30,
        style: undefined,
        color: '#000'
      }),
      undefined
    );
  });

  it('calls MaterialIcons with correct mapped names for chevron.left.forwardslash.chevron.right', () => {
    render(<IconSymbol name="chevron.left.forwardslash.chevron.right" color="#09f" size={32} />);
    expect(MaterialIcons).toHaveBeenCalledWith(
      expect.objectContaining({
        name: 'code',
        size: 32,
        style: undefined,
        color: '#09f'
      }),
      undefined
    );
  });

  it('calls MaterialIcons with correct mapped name for chevron.right', () => {
    render(<IconSymbol name="chevron.right" color="#123" size={20} />);
    expect(MaterialIcons).toHaveBeenCalledWith(
      expect.objectContaining({
        name: 'chevron-right',
        size: 20,
        style: undefined,
        color: '#123'
      }),
      undefined
    );
  });

  it('calls MaterialIcons with any custom style', () => {
    const customStyle = { margin: 14, opacity: 0.4 };
    render(<IconSymbol name="house.fill" color="#aab" style={customStyle} />);
    expect(MaterialIcons).toHaveBeenCalledWith(
      expect.objectContaining({
        name: 'home',
        size: 24,
        style: { margin: 14, opacity: 0.4 },
        color: '#aab'
      }),
      undefined
    );
  });

  it('calls MaterialIcons using default size and style when optional props are missing', () => {
    render(<IconSymbol name="paperplane.fill" color="#eee" />);
    expect(MaterialIcons).toHaveBeenCalledWith(
      expect.objectContaining({
        name: 'send',
        size: 24, // default size
        style: undefined,
        color: '#eee'
      }),
      undefined
    );
  });

  it('matches snapshot', () => {
    const tree = render(<IconSymbol name="house.fill" color="#000" />).toJSON();
    expect(tree).toMatchSnapshot();
  });
  
});

