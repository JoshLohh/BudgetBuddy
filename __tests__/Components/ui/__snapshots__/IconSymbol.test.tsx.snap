// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IconSymbol tests using toHaveBeenCalledWith pattern matches snapshot 1`] = `null`;
